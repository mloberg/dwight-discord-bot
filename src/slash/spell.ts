import { isNumber, sample } from 'lodash';

import spellList from '../data/spells';
import { FriendlyError } from '../error';
import { SlashCommand } from '../types';

const spell: SlashCommand = {
    name: 'spell',
    description: 'Return a random spell',
    options: [
        {
            name: 'level',
            description: 'Spell level',
            type: 'INTEGER',
            choices: [
                {
                    name: 'Cantrip',
                    value: 0,
                },
                {
                    name: '1st',
                    value: 1,
                },
                {
                    name: '2nd',
                    value: 2,
                },
                {
                    name: '3rd',
                    value: 3,
                },
                {
                    name: '4th',
                    value: 4,
                },
                {
                    name: '5th',
                    value: 5,
                },
                {
                    name: '6th',
                    value: 6,
                },
                {
                    name: '7th',
                    value: 7,
                },
                {
                    name: '8th',
                    value: 8,
                },
                {
                    name: '9th',
                    value: 9,
                },
            ],
        },
        {
            name: 'school',
            description: 'Spell school',
            type: 'STRING',
            choices: [
                {
                    name: 'Abjuration',
                    value: 'abjuration',
                },
                {
                    name: 'Conjuration',
                    value: 'conjuration',
                },
                {
                    name: 'Divination',
                    value: 'divination',
                },
                {
                    name: 'Enchanment',
                    value: 'enchantment',
                },
                {
                    name: 'Evocation',
                    value: 'evocation',
                },
                {
                    name: 'Illusion',
                    value: 'illusion',
                },
                {
                    name: 'Necromancy',
                    value: 'necromancy',
                },
                {
                    name: 'Transmutation',
                    value: 'transmutation',
                },
            ],
        },
        {
            name: 'class',
            description: 'Spell class',
            type: 'STRING',
            choices: [
                {
                    name: 'Bard',
                    value: 'Bard',
                },
                {
                    name: 'Cleric',
                    value: 'Cleric',
                },
                {
                    name: 'Druid',
                    value: 'Druid',
                },
                {
                    name: 'Paladin',
                    value: 'Paladin',
                },
                {
                    name: 'Ranger',
                    value: 'Ranger',
                },
                {
                    name: 'Warlock',
                    value: 'Warlock',
                },
                {
                    name: 'Wizard',
                    value: 'Wizard',
                },
            ],
        },
    ],
    async run(command) {
        const level = command.options.getInteger('level');
        const school = command.options.getString('school');
        const class_ = command.options.getString('class');

        let spells = await spellList();
        if (isNumber(level)) {
            spells = spells.filter((s) => s.level === level);
        }
        if (school) {
            spells = spells.filter((s) => s.school.toLowerCase() === school);
        }
        if (class_) {
            spells = spells.filter((s) => s.class.includes(class_));
        }

        const result = sample(spells);
        if (!result) {
            throw new FriendlyError("I couldn't find a spell matching those parameters.");
        }

        await command.reply(result.spell);
    },
};

export default spell;
