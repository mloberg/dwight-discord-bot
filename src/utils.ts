export function between(min: number, max: number): number {
    return Math.floor(Math.random()*(max-min+1)+min);
}

export function env(key: string, _default: string): string {
    return process.env[key] || _default;
}

export function rand<T>(items: T[]): T {
    return items[Math.floor(Math.random() * items.length)];
}

export function ucfirst(input: string): string {
    return input.charAt(0).toUpperCase() + input.slice(1).toLowerCase();
}

export function roll(expr: string): number {
    const input = expr.toLowerCase().match(/^(\d+)?d(\d+)([+-×x])?(\d+)?/);
    if (!input) {
        throw new Error("Invalid roll expression");
    }

    const count = Number(input[1] || 1);
    const dice = Number(input[2]);
    const modifier = input[3];
    const mod = Number(input[4]);

    let total = 0;
    for (let index = 0; index < count; index++) {
        total += Math.floor(Math.random() * (dice - 1 + 1)) + 1;
    }

    if ("+" === modifier) {
        total += mod;
    } else if ("-" === modifier) {
        total -= mod;
    } else if ("x" === modifier || "×" === modifier) {
        total = total * mod;
    }

    return total;
}
