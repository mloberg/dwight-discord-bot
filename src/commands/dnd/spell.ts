import { Command, CommandMessage, FriendlyError } from "discord.js-commando";

import spellList from "../../data/spells";
import { rand } from "../../utils";


export default class Treasure extends Command {
    constructor(client) {
        super(client, {
            name: "spell",
            aliases: ["scroll", "spellscroll"],
            group: "dnd",
            memberName: "spell",
            description: "Return a random spell.",
            examples: ["spell 0", "spell 1 Wizard Evocation"],
            format: "LEVEL CLASS SCHOOL",
            args: [
                {
                    key: "level",
                    prompt: "What level spell to return?",
                    type: "integer",
                },
                {
                    key: "class",
                    prompt: "What caster class should I filter by?",
                    type: "string",
                    default: "",
                    validate: c => ["bard", "cleric", "druid", "paladin", "ranger", "sorcerer", "warlock", "wizard"].includes(c.toLowerCase()),
                    parse: c => c.charAt(0).toUpperCase() + c.slice(1).toLowerCase(),
                },
                {
                    key: "school",
                    prompt: "What school of magic should I filter by?",
                    type: "string",
                    default: "",
                    validate: s => ["abjuration", "conjuration", "divination", "enchantment", "evocation", "illusion", "necromancy", "transmutation"].includes(s.toLowerCase()),
                    parse: s => s.toLowerCase(),
                }
            ],
        });
    }

    async run(msg: CommandMessage, args) {
        let spells = (await spellList()).filter(s => s.level === args.level);

        if (args.class) {
            spells = spells.filter(s => s.class.includes(args.class));
        }

        if (args.school) {
            spells = spells.filter(s => s.school === args.school);
        }

        const spell = rand(spells);
        if (!spell) {
            throw new FriendlyError("I couldn't find a spell with those filters.");
        }

        return msg.reply(`${spell.spell}`);
    }
}
