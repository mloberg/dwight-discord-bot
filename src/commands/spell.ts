import { Message, Client } from "discord.js";
import { Command, Arguments } from "../types";

import spellList from "../data/spells";
import { rand, ucfirst } from "../utils";

export default class extends Command {
    constructor(client: Client) {
        super(client, {
            name: "spell",
            description: "Return a random spell",
        });
    }

    async run(message: Message, args: Arguments) {
        const level = args.level || args._[0];
        const type = args.class || args._[1];
        const school = args.school || args._[2];

        let spells = await spellList();
        if (level) {
            spells = spells.filter(s => level === s.level || (s.level === 0 && level === 'cantrip'));
        }

        if (type) {
            spells = spells.filter(s => s.class.includes(ucfirst(type)));
        }

        if (school) {
            spells = spells.filter(s => school === s.school);
        }

        const spell = rand(spells);
        if (!spell) {
            throw new Error("I could not find an spell matching those parameters.");
        }

        return message.reply(spell.spell);
    }
}
