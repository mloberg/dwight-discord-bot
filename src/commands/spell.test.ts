import { CommandInteraction } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import { FriendlyError } from '../error';
import spell from './spell';

jest.mock('discord.js', () => ({
    CommandInteraction: jest.fn().mockImplementation(() => ({
        reply: jest.fn(),
        options: {
            getString: jest.fn(),
            getInteger: jest.fn(),
        },
    })),
}));
jest.mock('../data/spells', () => {
    return () => [
        {
            spell: 'Debugger',
            level: 1,
            school: 'Evocation',
            class: ['Wizard'],
        },
        {
            spell: 'Foo',
            level: 0,
            school: 'Illusion',
            class: ['Warlock', 'Ranger'],
        },
        {
            spell: 'Bar',
            level: 5,
            school: 'Divination',
            class: ['Bard', 'Cleric'],
        },
    ];
});

describe('/spell', () => {
    it('is a slash command', () => {
        expect(spell).toMatchSnapshot();
    });

    it('returns a random spell', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never));

        await spell.handler(command);
        expect(['Debugger', 'Foo', 'Bar']).toContainEqual(command.reply.mock.calls[0][0]);
    });

    it('returns a spell by level', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getInteger.mockReturnValue(0);

        await spell.handler(command);
        expect(command.reply).toHaveBeenCalledWith('Foo');
    });

    it('returns a spell matching multiple parameters', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getString.mockReturnValueOnce('evocation');
        command.options.getString.mockReturnValueOnce('Wizard');

        await spell.handler(command);
        expect(command.reply).toHaveBeenCalledWith('Debugger');
    });

    it('throws an error if no spell matches', async () => {
        const command = mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getString.mockReturnValueOnce('necromancy');
        command.options.getString.mockReturnValueOnce('Ranger');

        await expect(spell.handler(command)).rejects.toMatchError(
            new FriendlyError("I couldn't find a spell matching those parameters."),
        );
    });
});
