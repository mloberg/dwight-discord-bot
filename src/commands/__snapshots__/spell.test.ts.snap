// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/spell is a slash command 1`] = `
Object {
  "default_permission": undefined,
  "description": "Return a random spell",
  "name": "spell",
  "options": Array [
    Object {
      "choices": Array [
        Object {
          "name": "Cantrip",
          "value": 0,
        },
        Object {
          "name": "1st",
          "value": 1,
        },
        Object {
          "name": "2nd",
          "value": 2,
        },
        Object {
          "name": "3rd",
          "value": 3,
        },
        Object {
          "name": "4th",
          "value": 4,
        },
        Object {
          "name": "5th",
          "value": 5,
        },
        Object {
          "name": "6th",
          "value": 6,
        },
        Object {
          "name": "7th",
          "value": 7,
        },
        Object {
          "name": "8th",
          "value": 8,
        },
        Object {
          "name": "9th",
          "value": 9,
        },
      ],
      "description": "Spell level",
      "name": "level",
      "required": false,
      "type": 4,
    },
    Object {
      "choices": Array [
        Object {
          "name": "Abjuration",
          "value": "abjuration",
        },
        Object {
          "name": "Conjuration",
          "value": "conjuration",
        },
        Object {
          "name": "Divination",
          "value": "divination",
        },
        Object {
          "name": "Enchanment",
          "value": "enchantment",
        },
        Object {
          "name": "Evocation",
          "value": "evocation",
        },
        Object {
          "name": "Illusion",
          "value": "illusion",
        },
        Object {
          "name": "Necromancy",
          "value": "necromancy",
        },
        Object {
          "name": "Transmutation",
          "value": "transmutation",
        },
      ],
      "description": "Spell school",
      "name": "school",
      "required": false,
      "type": 3,
    },
    Object {
      "choices": Array [
        Object {
          "name": "Bard",
          "value": "Bard",
        },
        Object {
          "name": "Cleric",
          "value": "Cleric",
        },
        Object {
          "name": "Druid",
          "value": "Druid",
        },
        Object {
          "name": "Paladin",
          "value": "Paladin",
        },
        Object {
          "name": "Ranger",
          "value": "Ranger",
        },
        Object {
          "name": "Warlock",
          "value": "Warlock",
        },
        Object {
          "name": "Wizard",
          "value": "Wizard",
        },
      ],
      "description": "Spell class",
      "name": "class",
      "required": false,
      "type": 3,
    },
  ],
}
`;
